datanode/src/authorization.py

```diff
@@ -0,0 +1,361 @@
+
+    Returns:
+      USS identification from OAuth client_id or sub field
+
+    Raises:
+      HTTPException: when the access token is invalid or inappropriate
+    """
+    uss_id = None
+    if self.test_id:
+      if self.test_id in headers.get('access_token', ''):
+        return headers['access_token']
+      elif self.test_id in headers.get('Authorization', ''):
+        return headers['Authorization']
+      elif 'access_token' not in headers and 'Authorization' not in headers:
+        return self.test_id
+
+    # TODO(hikevin): Replace with OAuth Discovery and JKWS
+    token = None
+    if 'Authorization' in headers:
+      token = headers['Authorization'].replace('Bearer ', '')
+    elif 'access_token' in headers:
+      token = headers['access_token']
+    if not token:
+      log.error('Attempt to access resource without access_token in header.')
+      flask.abort(status.HTTP_403_FORBIDDEN,
+                  'Valid OAuth access_token must be provided in header.')
+
+    # Verify validity of claims without checking signature
+    try:
+      r = jwt.decode(token, algorithms='RS256', verify=False)
+      #TODO(hikevin): Check scope is valid for InterUSS Platform
+      uss_id = r['client_id'] if 'client_id' in r else r.get('sub', None)
+    except jwt.ImmatureSignatureError:
+      log.error('Access token is immature.')
+      flask.abort(status.HTTP_401_UNAUTHORIZED,
+                  'OAuth access_token is invalid: token is immature.')
+    except jwt.ExpiredSignatureError:
+      log.error('Access token has expired.')
+      flask.abort(status.HTTP_401_UNAUTHORIZED,
+                  'OAuth access_token is invalid: token has expired.')
+    except jwt.DecodeError:
+      log.error('Access token is invalid and cannot be decoded.')
+      flask.abort(status.HTTP_400_BAD_REQUEST,
+                  'OAuth access_token is invalid: token cannot be decoded.')
+    except jwt.InvalidTokenError as e:
+      log.error('Unexpected InvalidTokenError: %s', str(e))
+      flask.abort(status.HTTP_500_INTERNAL_SERVER_ERROR,
+                  'Unexpected token error: ' + str(e))
+    issuer = r.get('iss', None)
+
+    if tiles:
+      # Check only authorities that manage all specified tiles
+      authorities = set.intersection(
+          *[self._GetAuthorities(issuer, tile) for tile in tiles])
+    else:
+      authorities = self._GetAuthorities(issuer, None)
+    if not authorities:
+      flask.abort(status.HTTP_401_UNAUTHORIZED,
+                  'No authorization authorities could be found')
+
+    # Check signature against all possible public keys
+    valid = False
+    for authority in authorities:
+      try:
+        jwt.decode(token, authority.public_key, algorithms='RS256')
+        valid = True
+        break
+      except jwt.InvalidSignatureError:
+        # Access token signature not valid for this public key, but might be
+        # valid for a different public key.
+        pass
+    if not valid:
```

Comment:
140-164 is all for an invalid token? Do we need all of this different logic for different error messages, or can we just say the key is not valid? Just maintenance issue.

Reply:
Correct, 140-164 is all just to provide an informative error message to the client.  Some of the largest maintenance burdens for the InterUSS Platform so far have been unclear or misleading error messages (leading to difficult troubleshooting with partners; see, e.g., #73), so I think it's worth specifying exactly why a token has failed to validate.